src/Sarif/Visitors/InsertOptionalDataVisitor.cs

```diff
@@ -0,0 +1,148 @@
+
+        private Run _run;
+        private bool _overwriteExistingData;
+        private FileRegionsCache _fileRegionsCache;
+        private OptionallyEmittedData _dataToInsert;
+        
+        public InsertOptionalDataVisitor(OptionallyEmittedData dataToInsert, bool overwriteExistingData = true)
+        {
+            _dataToInsert = dataToInsert;
+            _overwriteExistingData = dataToInsert.Includes(OptionallyEmittedData.OverwriteExistingData);
+        }
+
+        public override Run VisitRun(Run node)
+        {
+            _run = node;
+
+            if (node != null)
+            {
+                bool scrapeFileReferences = _dataToInsert.Includes(OptionallyEmittedData.Hashes)      ||
+                                            _dataToInsert.Includes(OptionallyEmittedData.TextFiles)   ||
+                                            _dataToInsert.Includes(OptionallyEmittedData.BinaryFiles);
+
+                if (scrapeFileReferences)
+                {
+                    var visitor = new AddFileReferencesVisitor();
+                    visitor.VisitRun(node);
+                }
+
+                if (node.Files != null)
+                {
+                    var keys = node.Files.Keys.ToArray();
+                    foreach (var key in keys)
+                    {
+                        var value = node.Files[key];
+                        if (value != null)
+                        {
+                            node.Files[key] = VisitDictionaryValueNullChecked(key, value);
+                        }
+                    }
+                }
+            }
+
+            if (node.Results != null)
+            {
+                var results = new List<Result>();
+
+                foreach (Result result in node.Results)
+                {
+                    results.Add(VisitResult(result));
+                }
+
+                node.Results = results;
+            }
+            return node;
+        }
+
+        public override PhysicalLocation VisitPhysicalLocation(PhysicalLocation node)
+        {
+            if (node.Region == null || node.Region.IsBinaryRegion)
+            {
+                goto Exit;
+            }
+
+            bool insertRegionSnippets = _dataToInsert.Includes(OptionallyEmittedData.RegionSnippets);
+            bool insertContextCodeSnippets = _dataToInsert.Includes(OptionallyEmittedData.ContextCodeSnippets);
+            bool populateRegionProperties = _dataToInsert.Includes(OptionallyEmittedData.ComprehensiveRegionProperties);
+
+            if (insertRegionSnippets || populateRegionProperties)
+            {
+                Region result;
+
+                _fileRegionsCache = _fileRegionsCache ?? new FileRegionsCache(_run);
+                
+                // If we can resolve a file location to a newly constructed
+                // absolute URI, we will prefer that
+                if (!_run.TryResolveUri(node.FileLocation, out Uri resolvedUri))
```

Comment:
>_run.TryResolveUri [](start = 21, length = 18)

Remember my comment wondering whether it was right for you to populate the `out` parameter if the `Try` routine returns false? And I wondered if it violated "least surprise" for you to do so?

Well apparently it did, because here you surprised even yourself and re-filled `resolvedUri` with the exact value that your `Try` routine did :-)

So I think you _should_ change that initialization in `TryResolveUri`.

Reply:
Did that. Also changed helper name to 'TryReconstructAbsoluteUri'. returns true if we can reconstruct an absolute uri or if the input URI was already absolute. 

---
In reply to: [205985485](https://github.com/Microsoft/sarif-sdk/pull/974#discussion_r205985485) [](ancestors = 205985485)